src/14_legendre.jl

```diff
@@ -0,0 +1,372 @@
+
+    # step 1: compute Q_n^0(x)
+    q_n_0 = legendreq(n, x)
+
+    # step 2: compute Q_n^1(x)
+    Q01 = -(1-x^2)^(-0.5)                                                   # = Q_0^1(x)
+    Q11 = -sqrt(1-x^2) * (0.5*log((1+x)/(1-x)) + x/(1-x^2))                 # = Q_1^1(x)
+    if n == 0
+        q = Q01
+    elseif n == 1
+        q = Q11
+    else
+        q, q_prev = Q11, Q01
+        for k = 2:n
+            q_prev, q_prev_prev = q, q_prev
+            q = (2k-1)//(k-1) * x * q_prev - k//(k-1) * q_prev_prev         # = Q_k^1(x)
+        end                                                                 # on output: q = Q_n^1(x)
+    end
+    q_n_1 = q
+
+    # step 3: compute Q_n^m(x)
+    q_prev  = q_n_0
+    q       = q_n_1
+    for k = 2:m
+        q_prev, q_prev_prev = q, q_prev
+        q = -2*(k-1)*x*q_prev / sqrt(1-x^2) - (n+k-1)*(n-k+2)*q_prev_prev   # = Q_n^k(x)
+    end                                                                     # on output: q = Q_n^m(x)
+
+    q
+end
+
+@doc raw"""
+    ABC_recurrence(n, x, A, B, C, p0, p1)
+
+Evaluate polynomials by the recurrence
+```math
+p_n(x)
+=
+\begin{cases}
+    p_0(x)
+    & \text{for} \quad n = 0
+    \\
+    p_1(x)
+    & \text{for} \quad n = 1
+    \\
+    (A_n x + B_n) p_{n-1}(x) - C_n p_{n-2}(x)
+    & \text{for} \quad n \geq 2
+    \,.
+\end{cases}
+```
+
+This will be used to compute Legendre ``P_n``, Laguerre ``L_n``, Hermite ``H_n``,
+and Chebyshev polynomials of 1st kind ``T_n`` and 2nd kind ``U_n``.
+
+# Arguments
+- `n::Integer`:                     Polynomial order
+- `x::Real`:                        evaluation point
+- `A::Union{Integer,Function}`:     recurrence coefficient ``A_m``
+- `B::Union{Integer,Function}`:     recurrence coefficient ``B_m``
+- `C::Union{Integer,Function}`:     recurrence coefficient ``C_m``
+- `p0::Function`:                   function of degree 0, i.e. ``p_0``
+- `p1::Function`:                   function of degree 1, i.e. ``p_1``
+
+#Implementation
+For some polynomials the coefficients ``A_m, B_m, C_m`` are independent of ``m`` and for
+others they aren't. Which is why we use `Union{Integer,Function}` and functions can be
+easily created by anonymous functions.
+
+Degree 0, and 1 are supplied as a function. Efficiency is not an issue as it is only invoked
+once.
+
+Type stability:
+- `A,B,C` are either functions returning a value of same type as `x` or an integer which will
+be casted to the type of `x`. Thus, variables `x,Am,Bm,Cm` are all of the same type.
+- `p0,p1` are functions which have the same return type as the type of `x`.
+- Thus, all variables: `p_prev_prev,p_prev,p,Am,Bm,Cm` are always of type `x`.
+
+# Reference
+> JIN, J. M.;
+> JJIE, Zhang Shan.
+> "Computation of special functions".
+> Wiley, 1996.
+> (p. 23)
+"""
+function ABC_recurrence(n::Integer, x,
+        A::Union{Integer,Function}, B::Union{Integer,Function}, C::Union{Integer,Function},
+        p0::Function, p1::Function)
```

Comment:
These function arguments won't be inlined if they are not declared as parameters (e.g. `p0::P0` with a `where {P0}`) at the end.

Reply:
parameters `p0,p1` are now supplied as values of same type as `x`. this seems simpler and favorable to the previous way of supplying them as functions. only once evaluated anyways...
type info supplied as parametric method. in the docs about [parametric methods](https://docs.julialang.org/en/v1/manual/methods/#Parametric-Methods-1) it doesnt say anything about inling the parameters. where can I read more about it?
see commit [790fa5d](https://github.com/JuliaMath/SpecialFunctions.jl/pull/175/commits/790fa5daeae54a5a55a9356209a01e69c9431b05).

Reply:
You are still passing function parameters in a bunch of places, e.g. `hermiteh`.

The issue is that passing extremely simple functions like this gives a big performance hit unless the `ABC_recurrence` specializes on the particular function type at compile time, which allows the compiler to inline these simple functions.

By default, Julia's compilers do not specialize on `Function` argument types unless they are declared in explicit type parameters (`where` clauses), as explained here: https://docs.julialang.org/en/v1/manual/performance-tips/index.html#Be-aware-of-when-Julia-avoids-specializing-1

Reply:
replaced the argument functions for sequences ``A_m, B_m, C_m`` by a generator ``ABC``. this should be more efficient.
see commit [a95dbcd](https://github.com/JuliaMath/SpecialFunctions.jl/pull/175/commits/a95dbcdb7964bafba7b79a6dbf42fa38f0620b49).

Reply:
removed that function in commit [780b6df](https://github.com/JuliaMath/SpecialFunctions.jl/pull/175/commits/780b6df504afa8169b6529f46b78dcd1b9350ef6).